File: sdk/go/manual.go
// Go manual instrumentation — custom spans, attributes, and metrics.
//
// Use this AFTER Init() from auto.go has set up the global providers.
// auto.go configures otel.GetTracerProvider() and otel.GetMeterProvider();
// this file shows how to use them for custom business operations.

package telemetry

import (
	"context"
	"fmt"
	"time"

	"go.opentelemetry.io/otel"
	"go.opentelemetry.io/otel/attribute"
	"go.opentelemetry.io/otel/codes"
	"go.opentelemetry.io/otel/metric"
	"go.opentelemetry.io/otel/trace"
)

// ComponentTracer creates a tracer for a specific component.
// Call this once at package init time and reuse the tracer.
// The component name appears as the instrumentation scope in Tempo.
func ComponentTracer(name string) trace.Tracer {
	return otel.Tracer(name)
}

// ComponentMeter creates a meter for a specific component.
func ComponentMeter(name string) metric.Meter {
	return otel.Meter(name)
}

// Example: wrapping a business operation in a custom span.
// The tracer is typically a package-level variable created at init time.
var orderTracer = otel.Tracer("com.example.orders")

func ProcessOrder(ctx context.Context, orderID string, itemCount int) error {
	// Start a span. The span is linked to the parent span in ctx (if any).
	ctx, span := orderTracer.Start(ctx, "process-order",
		// SpanKindInternal: work happening inside the service, not an inbound/outbound call.
		trace.WithSpanKind(trace.SpanKindInternal),
	)
	defer span.End()

	// Set attributes. Use OTel semantic convention names where they exist.
	// https://opentelemetry.io/docs/specs/semconv/
	span.SetAttributes(
		attribute.String("order.id", orderID),
		attribute.Int("order.item_count", itemCount),
	)

	if err := doWork(ctx, orderID); err != nil {
		// RecordError attaches the error to the span with a stack trace.
		span.RecordError(err)
		span.SetStatus(codes.Error, err.Error())
		return fmt.Errorf("processing order %s: %w", orderID, err)
	}

	span.SetStatus(codes.Ok, "")
	return nil
}

// Example: adding span events (timestamped annotations on the span timeline).
func processPayment(ctx context.Context, amount float64) error {
	ctx, span := orderTracer.Start(ctx, "process-payment")
	defer span.End()

	// AddEvent: a timestamped log entry attached to this span.
	span.AddEvent("payment-gateway-called", trace.WithAttributes(
		attribute.String("gateway", "stripe"),
		attribute.Float64("amount.usd", amount),
	))

	// ... call payment gateway ...

	span.AddEvent("payment-gateway-response", trace.WithAttributes(
		attribute.String("gateway.status", "success"),
	))

	return nil
}

// Example: custom metrics alongside spans.
var (
	orderMeter    = otel.Meter("com.example.orders")
	// Counters and histograms are safe to use concurrently — create once, reuse.
	ordersCounter, _ = orderMeter.Int64Counter(
		"orders.processed",
		metric.WithDescription("Total orders processed"),
		metric.WithUnit("1"),
	)
	orderDuration, _ = orderMeter.Int64Histogram(
		"order.duration",
		metric.WithDescription("Time to process an order in milliseconds"),
		metric.WithUnit("ms"),
	)
)

func recordOrderMetrics(ctx context.Context, durationMs int64, success bool) {
	status := "success"
	if !success {
		status = "error"
	}

	// Counter: increment by 1 with labels.
	ordersCounter.Add(ctx, 1, metric.WithAttributes(
		attribute.String("status", status),
	))

	// Histogram: record the duration. Labels become Prometheus le buckets.
	orderDuration.Record(ctx, durationMs, metric.WithAttributes(
		attribute.String("status", status),
	))
}

// Example: propagating context across goroutines.
// Always pass ctx through your call stack — never call otel.Tracer().Start() with context.Background()
// inside a request handler, or the span will be disconnected from the trace.
func runAsync(ctx context.Context) {
	go func() {
		// Pass the parent context to preserve the trace chain.
		ctx, span := orderTracer.Start(ctx, "async-work")
		defer span.End()

		_ = doWork(ctx, "async")
		_ = time.Now() // placeholder
	}()
}

func doWork(ctx context.Context, id string) error {
	return nil
}

File: sdk/python/manual.py
"""
Python manual instrumentation — adding custom spans, attributes, and metrics.

Use this AFTER auto.py has initialized the TracerProvider and MeterProvider.
auto.py sets up the global providers; this module shows how to use them.
"""

from opentelemetry import trace, metrics
from opentelemetry.trace import SpanKind, StatusCode

# Get a tracer scoped to this component.
# The name appears as the instrumentation scope in Tempo's trace view.
tracer = trace.get_tracer("my-component", "1.0.0")

# Get a meter for recording custom metrics.
meter = metrics.get_meter("my-component", "1.0.0")

# Create a counter: records cumulative counts (orders processed, errors, cache misses).
orders_counter = meter.create_counter(
    "orders.processed",
    description="Total orders successfully processed",
    unit="1",
)

# Create a histogram: records distributions (latency, payload size, queue depth).
payment_duration = meter.create_histogram(
    "payment.duration",
    description="Time to process a payment in milliseconds",
    unit="ms",
)


def process_order(order_id: str, items: list) -> dict:
    """Wrap a business operation in a custom span."""
    # start_as_current_span: creates a span and makes it active in the current context.
    # Child spans created inside (by instrumented libraries) will be linked to this one.
    with tracer.start_as_current_span(
        "process-order",
        # kind: SERVER = inbound request, CLIENT = outbound call, INTERNAL = internal work.
        kind=SpanKind.INTERNAL,
    ) as span:
        # Set attributes using OTel semantic convention names where they exist.
        # Invented names are fine for domain-specific data.
        span.set_attribute("order.id", order_id)
        span.set_attribute("order.item_count", len(items))

        try:
            result = _do_work(order_id, items)

            # Record a metric alongside the span.
            orders_counter.add(1, {"status": "success", "region": "us-east-1"})
            span.set_status(StatusCode.OK)
            return result

        except Exception as e:
            # record_exception: attaches the stack trace to the span.
            span.record_exception(e)
            span.set_status(StatusCode.ERROR, str(e))
            orders_counter.add(1, {"status": "error"})
            raise


def charge_payment(amount_usd: float) -> bool:
    """Example: histogram measurement and span events."""
    import time

    with tracer.start_as_current_span("charge-payment") as span:
        span.set_attribute("payment.amount_usd", amount_usd)
        start = time.time()

        try:
            # Add a span event: a timestamped log entry attached to this span.
            span.add_event("payment-gateway-called", {"gateway": "stripe"})

            success = _call_payment_gateway(amount_usd)

            elapsed_ms = (time.time() - start) * 1000
            # Record to the histogram with attributes that become label dimensions in Prometheus.
            payment_duration.record(elapsed_ms, {"payment.status": "success"})

            return success
        except Exception as e:
            elapsed_ms = (time.time() - start) * 1000
            payment_duration.record(elapsed_ms, {"payment.status": "error"})
            span.record_exception(e)
            span.set_status(StatusCode.ERROR, str(e))
            raise


def _do_work(order_id, items):
    return {"order_id": order_id, "status": "processed"}


def _call_payment_gateway(amount):
    return True
